blockstore/src/in_memory_blockstore.rs

```diff
@@ -0,0 +1,268 @@
+use cid::CidGeneric;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use multihash::Multihash;
+
+use crate::{Blockstore, BlockstoreError, Result};
+
+/// Simple in-memory blockstore implementation.
+pub struct InMemoryBlockstore<const MAX_MULTIHASH_SIZE: usize> {
+    map: DashMap<CidGeneric<MAX_MULTIHASH_SIZE>, Vec<u8>>,
+}
+
+impl<const MAX_MULTIHASH_SIZE: usize> InMemoryBlockstore<MAX_MULTIHASH_SIZE> {
+    /// Create new empty in-memory blockstore
+    pub fn new() -> Self {
+        InMemoryBlockstore {
+            map: DashMap::new(),
+        }
+    }
+
+    fn get_cid(&self, cid: &CidGeneric<MAX_MULTIHASH_SIZE>) -> Result<Option<Vec<u8>>> {
+        Ok(self.map.get(cid).as_deref().cloned())
+    }
+
+    fn insert_cid(&self, cid: CidGeneric<MAX_MULTIHASH_SIZE>, data: &[u8]) -> Result<()> {
+        let cid_entry = self.map.entry(cid);
+        if matches!(cid_entry, Entry::Occupied(_)) {
+            return Err(BlockstoreError::CidExists);
+        }
+
+        cid_entry.insert(data.to_vec());
+
+        Ok(())
+    }
+}
+
+#[cfg_attr(not(docs_rs), async_trait::async_trait)]
+impl<const MAX_MULTIHASH_SIZE: usize> Blockstore for InMemoryBlockstore<MAX_MULTIHASH_SIZE> {
+    async fn get<const SS: usize>(&self, cid: &CidGeneric<SS>) -> Result<Option<Vec<u8>>> {
+        let hash = cid.hash();
+        let hash =
+            Multihash::wrap(hash.code(), hash.digest()).map_err(|_| BlockstoreError::CidTooLong)?;
+        let cid = CidGeneric::new_v1(cid.codec(), hash);
+
+        self.get_cid(&cid)
+    }
+
+    async fn put_keyed<const SS: usize>(&self, cid: &CidGeneric<SS>, data: &[u8]) -> Result<()> {
+        let hash = cid.hash();
+        let hash =
+            Multihash::wrap(hash.code(), hash.digest()).map_err(|_| BlockstoreError::CidTooLong)?;
+        let cid = CidGeneric::new_v1(cid.codec(), hash);
+
+        self.insert_cid(cid, data)
+    }
+}
```

Comment:
Implement also `Blockstore::has` method and use `DashMap::contains_key`. It will avoid allocation of the `get` result.

Reply:
added 👍